File: service/src/main/java/no/nav/melosys/service/brev/bestilling/HentMuligeProduserbaredokumenterService.java
package no.nav.melosys.service.brev.bestilling;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import no.nav.melosys.domain.Behandling;
import no.nav.melosys.domain.Fagsak;
import no.nav.melosys.domain.kodeverk.Aktoersroller;
import no.nav.melosys.domain.kodeverk.Mottakerroller;
import no.nav.melosys.domain.kodeverk.Sakstemaer;
import no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter;
import no.nav.melosys.exception.FunksjonellException;
import no.nav.melosys.service.behandling.BehandlingService;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import static java.util.Collections.emptyList;
import static no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter.*;

@Component
public class HentMuligeProduserbaredokumenterService {

    private final BehandlingService behandlingService;

    public HentMuligeProduserbaredokumenterService(BehandlingService behandlingService) {
        this.behandlingService = behandlingService;
    }

    @Transactional
    public List<Produserbaredokumenter> hentMuligeProduserbaredokumenter(long behandlingId, Mottakerroller mottakerroller) {
        Behandling behandling = behandlingService.hentBehandlingMedSaksopplysninger(behandlingId);
        Fagsak fagsak = behandling.getFagsak();

        if (behandling.erInaktiv()) {
            return emptyList();
        }

        return switch (mottakerroller) {
            case BRUKER -> {
                List<Produserbaredokumenter> brevmaler = new ArrayList<>();
                if (fagsak.getTema() == Sakstemaer.MEDLEMSKAP_LOVVALG && (behandling.erFørstegangsvurdering() || behandling.erNyVurdering())) {
                    brevmaler.add(MELDING_FORVENTET_SAKSBEHANDLINGSTID_SOKNAD);
                }
                if (!behandling.erManglendeInnbetalingTrygdeavgift()) {
                    brevmaler.add(MANGELBREV_BRUKER);
                }
                brevmaler.add(GENERELT_FRITEKSTBREV_BRUKER);
                yield brevmaler;
            }
            case VIRKSOMHET -> Collections.singletonList(GENERELT_FRITEKSTBREV_VIRKSOMHET);
            case ARBEIDSGIVER -> behandling.erManglendeInnbetalingTrygdeavgift()
                ? List.of(GENERELT_FRITEKSTBREV_ARBEIDSGIVER)
                : List.of(MANGELBREV_ARBEIDSGIVER, GENERELT_FRITEKSTBREV_ARBEIDSGIVER);
            case ANNEN_ORGANISASJON -> (fagsak.getHovedpartRolle() == Aktoersroller.VIRKSOMHET || behandling.erManglendeInnbetalingTrygdeavgift())
                ? Collections.singletonList(GENERELT_FRITEKSTBREV_VIRKSOMHET)
                : List.of(MANGELBREV_ARBEIDSGIVER, GENERELT_FRITEKSTBREV_ARBEIDSGIVER);
            case UTENLANDSK_TRYGDEMYNDIGHET -> Collections.singletonList(UTENLANDSK_TRYGDEMYNDIGHET_FRITEKSTBREV);
            case NORSK_MYNDIGHET -> Collections.singletonList(FRITEKSTBREV);
            default -> throw new FunksjonellException("Mottakerrollen " + mottakerroller + " kan ikke sende brev gjennom brevmenyen");
        };
    }
}

File: service/src/main/kotlin/no/nav/melosys/service/brev/bestilling/HentMuligeProduserbaredokumenterService.kt
package no.nav.melosys.service.brev.bestilling

import no.nav.melosys.domain.Behandling
import no.nav.melosys.domain.Fagsak
import no.nav.melosys.domain.kodeverk.Aktoersroller
import no.nav.melosys.domain.kodeverk.Mottakerroller
import no.nav.melosys.domain.kodeverk.Sakstemaer
import no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter
import no.nav.melosys.exception.FunksjonellException
import no.nav.melosys.service.behandling.BehandlingService
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class HentMuligeProduserbaredokumenterService(private val behandlingService: BehandlingService, private val unleash: Unleash) {

    @Transactional
    fun hentMuligeProduserbaredokumenter(behandlingId: Long, mottakerroller: Mottakerroller): List<Produserbaredokumenter> {
        val behandling = behandlingService.hentBehandlingMedSaksopplysninger(behandlingId)
        val fagsak = behandling.fagsak
        return if (behandling.erInaktiv()) {
            emptyList()
        } else when (mottakerroller) {
            Mottakerroller.BRUKER -> hentMuligeForBruker(behandling, behandling.fagsak)
            Mottakerroller.ARBEIDSGIVER ->
                if (behandling.erManglendeInnbetalingTrygdeavgift())
                    listOf(Produserbaredokumenter.GENERELT_FRITEKSTBREV_ARBEIDSGIVER)
                else
                    listOf(
                        Produserbaredokumenter.MANGELBREV_ARBEIDSGIVER,
                        Produserbaredokumenter.GENERELT_FRITEKSTBREV_ARBEIDSGIVER
                    )

            Mottakerroller.ANNEN_ORGANISASJON ->
                if (fagsak.hovedpartRolle == Aktoersroller.VIRKSOMHET || behandling.erManglendeInnbetalingTrygdeavgift())
                    listOf(Produserbaredokumenter.GENERELT_FRITEKSTBREV_VIRKSOMHET)
                else
                    listOf(
                        Produserbaredokumenter.MANGELBREV_ARBEIDSGIVER,
                        Produserbaredokumenter.GENERELT_FRITEKSTBREV_ARBEIDSGIVER
                    )

            Mottakerroller.VIRKSOMHET -> listOf(Produserbaredokumenter.GENERELT_FRITEKSTBREV_VIRKSOMHET)
            Mottakerroller.UTENLANDSK_TRYGDEMYNDIGHET -> listOf(Produserbaredokumenter.UTENLANDSK_TRYGDEMYNDIGHET_FRITEKSTBREV)
            Mottakerroller.NORSK_MYNDIGHET -> listOf(Produserbaredokumenter.FRITEKSTBREV)
            else -> throw FunksjonellException("Mottakerrollen $mottakerroller kan ikke sende brev gjennom brevmenyen")
        }
    }

    private fun hentMuligeForBruker(behandling: Behandling, fagsak: Fagsak): List<Produserbaredokumenter> {
        if (behandling.erManglendeInnbetalingTrygdeavgift()) {
            return listOf(Produserbaredokumenter.GENERELT_FRITEKSTBREV_BRUKER)
        }

        val muligeProduserbareDokumenter = hentDefaultMuligeForBruker()

        if (behandling.erÅrsavregning()) {
            muligeProduserbareDokumenter.add(Produserbaredokumenter.INNHENTING_AV_INNTEKTSOPPLYSNINGER)
        }

        if (fagsak.tema == Sakstemaer.MEDLEMSKAP_LOVVALG && (behandling.erFørstegangsvurdering() || behandling.erAndregangsbehandling())) {
            return listOf(Produserbaredokumenter.MELDING_FORVENTET_SAKSBEHANDLINGSTID_SOKNAD) + muligeProduserbareDokumenter
        }

        return muligeProduserbareDokumenter
    }

    private fun hentDefaultMuligeForBruker() = mutableListOf(
        Produserbaredokumenter.MANGELBREV_BRUKER,
        Produserbaredokumenter.GENERELT_FRITEKSTBREV_BRUKER
    )
}
